protocol/v1/qbft/validation/changeround/validate_justification.go

```diff
@@ -52,49 +52,8 @@ func (p *validateJustification) Run(signedMessage *specqbft.SignedMessage) error
 	}
 
 	for _, rcj := range data.RoundChangeJustification {
-		if rcj.Message == nil {
-			return errors.New("change round justification msg is nil")
-		}
-		if rcj.Message.MsgType != specqbft.PrepareMsgType {
-			return errors.Errorf("change round justification msg type not Prepare (%d)", rcj.Message.MsgType)
-		}
-		if signedMessage.Message.Height != rcj.Message.Height {
-			return errors.New("change round justification sequence is wrong")
-		}
-		if signedMessage.Message.Round <= rcj.Message.Round {
-			return errors.New("round change justification invalid: msg round wrong")
-		}
-		if data.PreparedRound != rcj.Message.Round {
-			return errors.New("round change justification invalid: msg round wrong")
-		}
-		if !bytes.Equal(signedMessage.Message.Identifier, rcj.Message.Identifier) {
-			return errors.New("change round justification msg Lambda not equal to msg Lambda not equal to instance lambda")
-		}
-		prepareMsg, err := rcj.Message.GetPrepareData()
-		if err != nil {
-			return errors.Wrap(err, "failed to get prepare data")
-		}
-		if !bytes.Equal(data.PreparedValue, prepareMsg.Data) {
-			return errors.New("round change justification invalid: prepare data != proposed data")
-		}
-		if len(rcj.GetSigners()) != 1 {
-			return errors.New("round change msg allows 1 signer")
-		}
-
-		// validateJustification justification signature
-		pksMap, err := p.share.PubKeysByID(rcj.GetSigners())
-		var pks beacon.PubKeys
-		for _, v := range pksMap {
-			pks = append(pks, v)
-		}
-
-		if err != nil {
-			return errors.Wrap(err, "change round could not get pubkey")
-		}
-		aggregated := pks.Aggregate()
-
-		if err = rcj.Signature.Verify(rcj, types.GetDefaultDomain(), spectypes.QBFTSignatureType, aggregated.Serialize()); err != nil {
-			return errors.Wrap(err, "change round could not verify signature")
+		if err := p.validateRoundChangeJustification(rcj, data, signedMessage); err != nil {
+			return fmt.Errorf("round change justification invalid: %w", err)
 		}
 	}
 
```

Comment:
@nkryuchkov 
Regarding - `protocol/v1/qbft/validation/changeround/validate_justification.go:128`

this quorum check is from spec?

Reply:
@nivBlox Not exactly. Here is how the check looked like in ssv:
```go
func (i *Instance) changeRoundQuorum(msgs []*specqbft.SignedMessage) (quorum bool, t int, n int) {
	quorum = len(msgs)*3 >= i.ValidatorShare.CommitteeSize()*2
	return quorum, len(msgs), i.ValidatorShare.CommitteeSize()
}
```

In spec it looks like this:
```go
// HasQuorum returns true if a unique set of signers has quorum
func HasQuorum(share *types.Share, msgs []*SignedMessage) bool {
	uniqueSigners := make(map[types.OperatorID]bool)
	for _, msg := range msgs {
		for _, signer := range msg.GetSigners() {
			uniqueSigners[signer] = true
		}
	}
	return share.HasQuorum(len(uniqueSigners))
}

func (share *Share) HasQuorum(cnt int) bool {
	return uint64(cnt) >= share.Quorum
}
```

So the most significant difference was checking length of messages instead of length of unique signers, so I added this check to `changeRoundQuorum` in ssv and renamed it to `HasQuorum`. There's another difference in quorum calculation, ssv does that with `len(msgs)*3 >= i.ValidatorShare.CommitteeSize()*2` but spec does this with `uint64(cnt) >= share.Quorum`, however, ssv's `Share` doesn't have the `Quorum` field, so it's impossible just to copy `HasQuorum` from spec, therefore I kind of merged it into what ssv had by checking unique signers length instead of messages length:

```go
func (i *Instance) changeRoundQuorum(msgs []*specqbft.SignedMessage) (quorum bool, t int, n int) {
	uniqueSigners := make(map[types.OperatorID]bool)
	for _, msg := range msgs {
		for _, signer := range msg.GetSigners() {
			uniqueSigners[signer] = true
		}
	}
	quorum = len(uniqueSigners)*3 >= i.ValidatorShare.CommitteeSize()*2
	return quorum, len(uniqueSigners), i.ValidatorShare.CommitteeSize()
}
```